<a href="https://colab.research.google.com/github/srgarcess/BCG/blob/main/task2/EDA_Cleaning_BCG.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



# **Exploratory Data Analysis and Cleaning**
By Godine Anvesh









## **The Datasets**

The dataset `ml_case_training_output.csv` named as `pco_output` contains:
> * `id`: contact id
*`churn`: has the client churned over the next 3 months


The dataset `ml_case_training_hist_data.csv` named as `pco_hist` contains the history of energy and power consumption per client:
>* `id`: contact id
*`price_date`: reference date
*`price_p1_var`: price of energy for the 1st period
*`price_p2_var`: price of energy for the 2nd period
*`price_p3_var`: price of energy for the 3rd period
*`price_p1_fix`: price of power for the 1st period
*`price_p2_fix`: price of power for the 2nd period
*`price_p3_fix`: price of power for the 3rd period


The dataset `ml_case_training_data.csv` contains:
>* `id`: contact id
* `activity_new`: category of the company's activity. 419 unique values, remove `NaN`
* `campaign_disc_elec`: code of the electricity campaign the customer last subscribed to. 0 non-null
* `channel_sales`: code of the sales channel
* `cons_12m`: electricity consumption of the past 12 months
* `cons_gas_12m`: gas consumption of the past 12 months
* `cons_last_month`: electricity consupmtion of the last month
* `date_activ`: date of activation of the contract
* `date_end`: registered date of the end of the contract
* `date_first_activ`: date of first contract of the client
* `date_modif_prod`: date of last modification of the product
*  `date_renewal`: date of the next contract renewal
* `forecast_base_bill_ele`: forecasted electricity bill baseline for next month
* `forecast_base_bill_year`: forecasted electricity bill baseline for calendar year
* `forecast_bill_12m`: forecasted electricity bill baseline for 12 months
* `forecast_cons`: forecasted electricity consumption for next month
* `forecast_cons_12m`: forecasted electricity consumption for next 12 months
* `forecast_cons_year`: forecasted electricity consumption for next calendar year
* `forecast_discount_energy`: forecasted value of current discount
* `forecast_meter_rent_12m`: forecasted bill of meter rental for the next 12 months
* `forecast_price_energy_p1`: forecasted energy price for 1st period
* `forecast_price_energy_p2`: forecasted energy price for 2nd period
* `forecast_price_pow_p1`: forecasted power price for 1st period
* `has_gas`: indicated if client is also a gas client
* `imp_cons`: current paid consumption
* `margin_gross_pow_ele`: gross margin on power subscription
* `margin_net_pow_ele`: net margin on power subscription
* `nb_prod_act`: number of active products and services
* `net_margin`: total net margin
* `num_years_antig`: antiquity of the client (in number of years)
* `origin_up`: code of the electricity campaign the customer first subscribed to
* `pow_max`: subscribed power

## **Importing Libraries and Datasets**

In this section we import the libraries of interest as well as the datasets.

In [12]:
# Import libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import missingno as msno
from scipy.stats import zscore as zscore

ModuleNotFoundError: No module named 'missingno'

In [ ]:
# Read in dataset
url = 'https://raw.githubusercontent.com/godineanvesh/BCG-Forage-Task/main/ml_case_training_data.csv'
url2 = 'https://raw.githubusercontent.com/godineanvesh/BCG-Forage-Task/main/ml_case-training_hist_data.csv'
url3 = 'https://raw.githubusercontent.com/godineanvesh/BCG-Forage-Task/main/ml_case_tarining_output.csv'
# list of dates
dt_lst = ['date_activ','date_end','date_first_activ','date_modif_prod','date_renewal']

# data importing          
pco_main = pd.read_csv(url, parse_dates=dt_lst)#, index_col= 'date_activ')
pco_hist = pd.read_csv(url2, parse_dates=['price_date']) # Yearly history of consumption per user
pco_output = pd.read_csv(url3)
pd.set_option('display.max_columns',None)

#**Data Exploration**

## **The Output Dataset**

From the output dataset we can derive a quick insight on customer retention. 

In [ ]:
# Replace the churn column with appropiate labels
pco_output['churn'] = pco_output['churn'].replace({0:'Stayed',1:'Churned'})

# Glimpse
pco_output.head()

In [ ]:
# What number of customers have churned in the last 3 months?
attrition_count = pco_output['churn'].value_counts()
print('Total Number of Churned Customers:\n', attrition_count)

In [ ]:
# What is the proportion of customer attrition in the last 3 months?
attrition_rate = pco_output['churn'].value_counts() / pco_output.shape[0] * 100
print('Attrition rate: \n', attrition_rate)

---
###**Facts**
> * In the last 3 months 1,595 customers have churned
* There are currently 14,501 active clients
* Customer retention is 90% in the last 3 months 
* Customer attrition is 10% in the last 3 months

---
###**Observations**
* Dataset has complete cases


## **The History Dataset**

This dataset contains 1-year historical data for each client. It provides insights of the yearly activity of each client. 

In [13]:
# Display the yearly consumption of energy and power of customers
pco_hist.head()

id price_date  price_p1_var  price_p2_var  \
0  038af19179925da21a25619c5a24b745 2015-01-01      0.151367           0.0   
1  038af19179925da21a25619c5a24b745 2015-01-02      0.151367           0.0   
2  038af19179925da21a25619c5a24b745 2015-01-03      0.151367           0.0   
3  038af19179925da21a25619c5a24b745 2015-01-04      0.149626           0.0   
4  038af19179925da21a25619c5a24b745 2015-01-05      0.149626           0.0   

   price_p3_var  price_p1_fix  price_p2_fix  price_p3_fix  
0           0.0     44.266931           0.0           0.0  
1           0.0     44.266931           0.0           0.0  
2           0.0     44.266931           0.0           0.0  
3           0.0     44.266931           0.0           0.0  
4           0.0     44.266931           0.0           0.0

In [14]:
# Examing the structure of the dataframe
pco_hist.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193002 entries, 0 to 193001
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   id            193002 non-null  object        
 1   price_date    193002 non-null  datetime64[ns]
 2   price_p1_var  191643 non-null  float64       
 3   price_p2_var  191643 non-null  float64       
 4   price_p3_var  191643 non-null  float64       
 5   price_p1_fix  191643 non-null  float64       
 6   price_p2_fix  191643 non-null  float64       
 7   price_p3_fix  191643 non-null  float64       
dtypes: datetime64[ns](1), float64(6), object(1)
memory usage: 11.8+ MB


In [15]:
# Examine the descriptive statistics of the dataframe
pco_hist.describe()

price_p1_var   price_p2_var   price_p3_var   price_p1_fix  \
count  191643.000000  191643.000000  191643.000000  191643.000000   
mean        0.140991       0.054412       0.030712      43.325546   
std         0.025117       0.050033       0.036335       5.437952   
min         0.000000       0.000000       0.000000      -0.177779   
25%         0.125976       0.000000       0.000000      40.728885   
50%         0.146033       0.085483       0.000000      44.266930   
75%         0.151635       0.101780       0.072558      44.444710   
max         0.280700       0.229788       0.114102      59.444710   

        price_p2_fix   price_p3_fix  
count  191643.000000  191643.000000  
mean       10.698201       6.455436  
std        12.856046       7.782279  
min        -0.097752      -0.065172  
25%         0.000000       0.000000  
50%         0.000000       0.000000  
75%        24.339581      16.226389  
max        36.490692      17.458221

In [16]:
# Identify the nullity of the dataframe
missing_values_hist = pco_hist.isna().sum()
print('Total Missing Values:\n', missing_values_hist)

Total Missing Values:
 id                 0
price_date         0
price_p1_var    1359
price_p2_var    1359
price_p3_var    1359
price_p1_fix    1359
price_p2_fix    1359
price_p3_fix    1359
dtype: int64


In [17]:
# Identify the percentage of nullity in the dataframe for each collumn
missing_values_hist_perc = pco_hist.isnull().mean() * 100
print('Percentage of Missing Values:\n', missing_values_hist_perc)

Percentage of Missing Values:
 id              0.000000
price_date      0.000000
price_p1_var    0.704138
price_p2_var    0.704138
price_p3_var    0.704138
price_p1_fix    0.704138
price_p2_fix    0.704138
price_p3_fix    0.704138
dtype: float64


# Facts


* The average price of energy for the 1st period was: $0.14
* The average price of energy for the 2nd period was: \$0.05
* The average price of energy for the 3rd period was: \$0.03

*The average price of energy was declining in the last year*.

* The average power of power for the 1st period was: \$43.32
* The average power of power for the 2nd period was: \$10.69
* The average power of power for the 3rd period was: \$6.45

*The average price of power was declining in the last year*.











# Observations

* The columns `price_p1_fix`, `price_p2_fix`, and `price_p3_fix` contain negative values. *These negative prices of power do not make sense.* 

* The dataset `pco_hist` contains 1359 rows displaying `NaN` values on 6 variables except for `id` and `price_date`.

* The `price_..._var` and `price_..._fix` columns are missing **0.704%** of the data in each of them.

Note: `Pandas` *recognizes these* `NaN` *values and removes them when displaying descriptives statistics*.

*Notice how the price of energy has a minimum value of zero. Perhaps some customers churned thereby making the consumption of energy zero.*

---

# **The Main Dataset**

This dataset contain more characteristics about each client's account and activity.

In [18]:
# Print header
pco_main.head()

id                      activity_new  \
0  48ada52261e7cf58715202705a0451c9  esoiiifxdlbkcsluxmfuacbdckommixw   
1  24011ae4ebbe3035111d65fa7c15bc57                               NaN   
2  d29c2c54acc38ff3c0614d0a653813dd                               NaN   
3  764c75f661154dac3a6c254cd082ea7d                               NaN   
4  bba03439a292a1e166f80264c16191cb                               NaN   

   campaign_disc_ele                     channel_sales  cons_12m  \
0                NaN  lmkebamcaaclubfxadlmueccxoimlema    309275   
1                NaN  foosdfpfkusacimwkcsosbicdxkicaua         0   
2                NaN                               NaN      4660   
3                NaN  foosdfpfkusacimwkcsosbicdxkicaua       544   
4                NaN  lmkebamcaaclubfxadlmueccxoimlema      1584   

   cons_gas_12m  cons_last_month date_activ   date_end date_first_activ  \
0             0            10025 2012-07-11 2016-06-11              NaT   
1         54946                0 2013-06-15 2016-06-15              NaT   
2             0                0 2009-08-21 2016-08-30              NaT   
3             0                0 2010-04-16 2016-04-16              NaT   
4             0                0 2010-03-30 2016-03-30              NaT   

  date_modif_prod date_renewal  forecast_base_bill_ele  \
0      2012-07-11   2015-09-11                     NaN   
1             NaT   2015-06-23                     NaN   
2      2009-08-21   2015-08-31                     NaN   
3      2010-04-16   2015-04-17                     NaN   
4      2010-03-30   2015-03-31                     NaN   

   forecast_base_bill_year  forecast_bill_12m  forecast_cons  \
0                      NaN                NaN            NaN   
1                      NaN                NaN            NaN   
2                      NaN                NaN            NaN   
3                      NaN                NaN            NaN   
4                      NaN                NaN            NaN   

   forecast_cons_12m  forecast_cons_year  forecast_discount_energy  \
0           26520.30               10025                       0.0   
1               0.00                   0                       0.0   
2             189.95                   0                       0.0   
3              47.96                   0                       0.0   
4             240.04                   0                       0.0   

   forecast_meter_rent_12m  forecast_price_energy_p1  \
0                   359.29                  0.095919   
1                     1.78                  0.114481   
2                    16.27                  0.145711   
3                    38.72                  0.165794   
4                    19.83                  0.146694   

   forecast_price_energy_p2  forecast_price_pow_p1 has_gas  imp_cons  \
0                  0.088347              58.995952       f     831.8   
1                  0.098142              40.606701       t       0.0   
2                  0.000000              44.311378       f       0.0   
3                  0.087899              44.311378       f       0.0   
4                  0.000000              44.311378       f       0.0   

   margin_gross_pow_ele  margin_net_pow_ele  nb_prod_act  net_margin  \
0                -41.76              -41.76            1     1732.36   
1                 25.44               25.44            2      678.99   
2                 16.38               16.38            1       18.89   
3                 28.60               28.60            1        6.60   
4                 30.22               30.22            1       25.46   

   num_years_antig                         origin_up  pow_max  
0                3  ldkssxwpmemidmecebumciepifcamkci  180.000  
1                3  lxidpiddsbxsbosboudacockeimpuepw   43.648  
2                6  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.800  
3                6  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.856  
4                6  kamkkxfxxuwbdslkwifmmcsiusiuosws   1

In [19]:
# print info
pco_main.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16096 entries, 0 to 16095
Data columns (total 32 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   id                        16096 non-null  object        
 1   activity_new              6551 non-null   object        
 2   campaign_disc_ele         0 non-null      float64       
 3   channel_sales             11878 non-null  object        
 4   cons_12m                  16096 non-null  int64         
 5   cons_gas_12m              16096 non-null  int64         
 6   cons_last_month           16096 non-null  int64         
 7   date_activ                16096 non-null  datetime64[ns]
 8   date_end                  16094 non-null  datetime64[ns]
 9   date_first_activ          3508 non-null   datetime64[ns]
 10  date_modif_prod           15939 non-null  datetime64[ns]
 11  date_renewal              16056 non-null  datetime64[ns]
 12  forecast_base_bill

In [20]:
# Identify the percentage of nullity in the dataframe for each collumn
missing_values_main_perc = pco_main.isnull().mean() * 100
print('Percentage of Missing Values:\n', missing_values_main_perc)

Percentage of Missing Values:
 id                            0.000000
activity_new                 59.300447
campaign_disc_ele           100.000000
channel_sales                26.205268
cons_12m                      0.000000
cons_gas_12m                  0.000000
cons_last_month               0.000000
date_activ                    0.000000
date_end                      0.012425
date_first_activ             78.205765
date_modif_prod               0.975398
date_renewal                  0.248509
forecast_base_bill_ele       78.205765
forecast_base_bill_year      78.205765
forecast_bill_12m            78.205765
forecast_cons                78.205765
forecast_cons_12m             0.000000
forecast_cons_year            0.000000
forecast_discount_energy      0.782803
forecast_meter_rent_12m       0.000000
forecast_price_energy_p1      0.782803
forecast_price_energy_p2      0.782803
forecast_price_pow_p1         0.782803
has_gas                       0.000000
imp_cons                      0.0

In [21]:
# Examine the descriptive statistics of the main dataset
pco_main.describe()

campaign_disc_ele      cons_12m  cons_gas_12m  cons_last_month  \
count                0.0  1.609600e+04  1.609600e+04     1.609600e+04   
mean                 NaN  1.948044e+05  3.191164e+04     1.946154e+04   
std                  NaN  6.795151e+05  1.775885e+05     8.235676e+04   
min                  NaN -1.252760e+05 -3.037000e+03    -9.138600e+04   
25%                  NaN  5.906250e+03  0.000000e+00     0.000000e+00   
50%                  NaN  1.533250e+04  0.000000e+00     9.010000e+02   
75%                  NaN  5.022150e+04  0.000000e+00     4.127000e+03   
max                  NaN  1.609711e+07  4.188440e+06     4.538720e+06   

       forecast_base_bill_ele  forecast_base_bill_year  forecast_bill_12m  \
count             3508.000000              3508.000000        3508.000000   
mean               335.843857               335.843857        3837.441866   
std                649.406000               649.406000        5425.744327   
min               -364.940000              -364.940000       -2503.480000   
25%                  0.000000                 0.000000        1158.175000   
50%                162.955000               162.955000        2187.230000   
75%                396.185000               396.185000        4246.555000   
max              12566.080000             12566.080000       81122.630000   

       forecast_cons  forecast_cons_12m  forecast_cons_year  \
count    3508.000000       16096.000000        16096.000000   
mean      206.845165        2370.555949         1907.347229   
std       455.634288        4035.085664         5257.364759   
min         0.000000      -16689.260000       -85627.000000   
25%         0.000000         513.230000            0.000000   
50%        42.215000        1179.160000          378.000000   
75%       228.117500        2692.077500         1994.250000   
max      9682.890000      103801.930000       175375.000000   

       forecast_discount_energy  forecast_meter_rent_12m  \
count              15970.000000             16096.000000   
mean                   0.991547                70.309945   
std                    5.160969                79.023251   
min                    0.000000              -242.960000   
25%                    0.000000                16.230000   
50%                    0.000000                19.440000   
75%                    0.000000               131.470000   
max                   50.000000              2411.690000   

       forecast_price_energy_p1  forecast_price_energy_p2  \
count              15970.000000              15970.000000   
mean                   0.135901                  0.052951   
std                    0.026252                  0.048617   
min                    0.000000                  0.000000   
25%                    0.115237                  0.000000   
50%                    0.142881                  0.086163   
75%                    0.146348                  0.098837   
max                    0.273963                  0.195975   

       forecast_price_pow_p1      imp_cons  margin_gross_pow_ele  \
count           15970.000000  16096.000000          16083.000000   
mean               43.533496    196.123447             22.462276   
std                 5.212252    494.366979             23.700883   
min                -0.122184  -9038.210000           -525.540000   
25%                40.606701      0.000000             11.960000   
50%                44.311378     44.465000             21.090000   
75%                44.311378    218.090000             29.640000   
max                59.444710  15042.790000            374.640000   

       margin_net_pow_ele   nb_prod_act    net_margin  num_years_antig  \
count        16083.000000  16096.000000  16081.000000     16096.000000   
mean            21.460318      1.347788    217.987028         5.030629   
std             27.917349      1.459808    366.742030         1.676101   
min           -615.660000      1.000000  -4148.990000         1.000000   
25%          

---
# **Facts**
* The average tenure of a client is 5 years
* The average net marging is $217

---
# **Observations**
* The 14 columns contain negative minimum values

* The `activity_new` column is missing 59.3% of its data
* The `campaign_disc_ele` column is missing completely
* The `channel_sales` column is missing 26.2% of its data
* The `date_end` column is missing 0.01% of its data
* The `date_first_activ_` column is missing 78.2% of its data
* The `date_modif_prod` column is missing 0.9% of its data
* The `date_renewal` column is missing 0.24% of it data
* The `marging_gross_pow_ele` and `margin_net_pow_ele` columns are both missing 0.08% of its data
* The `net_margin` column is missing 0.09% of its data
* The `origin_up` column is missing 0.54% of its data
* The `pow_max` column is missing 0.018% of its data
* The `forecast_base_bill_ele`, `forecast_base_bill_year`, `forecast_bill_12m`, and `forecast_cons` columns are each missing 78.2% of its data

# **Data Cleaning and Imputation**

## **Dealing with missing data**

**Workflow for treating missing values**

1. Convert all missing values to null values
2. Analyze the amount and type of missingness in the data
3. Appropriately delete or impute missing values
4. Evaluate & compare the performance of the treated/imputed dataset

The `missingno` (imported as `msno`) package is great for visualizing missing data - we will be using:

- `msno.matrix()` visualizes a missingness matrix
- `msno.bar()` visualizes a missngness barplot
- `plt.show()` to show the plot

# **Is the data missing at random?**

**Types of missingness**

1. Missing Completely at Random (MCAR)
> *Missingness has no relationship between any values, observed or missing*
2. Missing at Random (MAR)
> *There is a systematic relationship between missingness and other observed data, but not the missing data*
3. Missing Not at Random (MNAR)
> *There is a relationship between missingness and its values, missing or non-missing*

# **When and how to delete missing data?**

**Types of deletions**
1. Pairwise deletion
>Pandas skips `NaN` whic is equivalent to pairwise deletion. Pairwise deletions minimize the amount of data loss and are hence preferred. However, it is also true that at several instances they might negatively affect our analysis. 
2. Listwise deletion
> In listwise deletion the incomplete row is deleted, also called complete case analysis. The major disadvantage of listwise deletions is amount of data lost. 
>Example: `df.dropna(subset=['column'], how='any',inplace=True)`


Note: *Both of these deletions are used only when the values are missing completely at random that is MCAR*

# **The History Dataset**

In [22]:
# Identify negative columns
negative_cols = ['price_p1_fix','price_p2_fix','price_p3_fix']

# Convert to positive the negative columns in pco_hist
pco_hist[negative_cols] = pco_hist[negative_cols].apply(abs)

pco_hist.describe()

price_p1_var   price_p2_var   price_p3_var   price_p1_fix  \
count  191643.000000  191643.000000  191643.000000  191643.000000   
mean        0.140991       0.054412       0.030712      43.325563   
std         0.025117       0.050033       0.036335       5.437816   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.125976       0.000000       0.000000      40.728885   
50%         0.146033       0.085483       0.000000      44.266930   
75%         0.151635       0.101780       0.072558      44.444710   
max         0.280700       0.229788       0.114102      59.444710   

        price_p2_fix   price_p3_fix  
count  191643.000000  191643.000000  
mean       10.698210       6.455443  
std        12.856039       7.782273  
min         0.000000       0.000000  
25%         0.000000       0.000000  
50%         0.000000       0.000000  
75%        24.339581      16.226389  
max        36.490692      17.458221

### **Visualizing the amount of missingness**







In [23]:
# Visualize the completeness of the dataframe
msno.bar(pco_hist)

NameError: name 'msno' is not defined

To the untrained eye, it might seem that there's no data missing. However, we estimated that 0.7% of the data in the price columns are missing. We can notice that the value counts at the top of each columns display a different amount.

In [ ]:
# Visualize the locations of the missing values of the dataset
sorted = pco_hist.sort_values(by = ['id','price_date'])
msno.matrix(sorted)


The nullity matrix describes the nullity of the dataset and appears blank wherever there are missing values. 

The column on the very right summarizes the general shape of the data completeness and points out the row. Total count of columns at the bottom right.

In [ ]:
# Visualize the correlation between the numeric variables of the dataframe
msno.heatmap(pco_hist)

In [ ]:
# Identify the index of the IDs containing missing values.
hist_NAN_index = pco_hist[pco_hist.isnull().any(axis=1)].index.values.tolist()

# Obtain a dataframe with the missing values
pco_hist_missing = pco_hist.iloc[hist_NAN_index,:]

# Glimpse at the NaN cases of the pco_hist dataset
pco_hist_missing.head(10)

In [ ]:
# extract the unique dates of missing data
date_lst = pco_hist_missing['price_date'].unique()
id_lst = pco_hist_missing['id'].unique()

# Create a time dataframe with the unique dates
time_df = pd.DataFrame(data=date_lst, columns=['price_date'] )

# Glimpse the time dataframe
time_df.sort_values(by=['price_date'])

---
# **Facts**

* There is high correlation between the missingness in the numeric columns and is values, missing or non-missing
* There are 1359 clients who are missing price data at least in 1 month

---
# **Observations** 

* After sorting the `pco_hist` dataset by `id` and `price_date`, we found that some columns are likely to be **MNAR**.
* The columns containing prices display strong positive correlation  in the missingness suggests a case of **MNAR**. 
* This event suggest that multicolinearity might be present in the dataset. 

# **Imputations**

Imputing time-series data requires a specialized treatment. Time-series data usually comes with special characteristics such trend, seasonality and cyclicality of which we can exploit when imputing missing values in the data.

In this particular dataset, there's not such thing as seasonality because it only has monthly data for one year.

# **Filling Time Series Data**

In [ ]:
# Make a copy of pco_hist dataset
pco_hist_ff = pco_hist.copy(deep=True)

# Print prior to imputing missing values
print(pco_hist_ff.iloc[hist_NAN_index,3:9].head())

# Fill NaNs using forward fill
pco_hist_ff.fillna(method = 'ffill', inplace=True)

print(pco_hist_ff.iloc[hist_NAN_index,3:9].head())


In [ ]:
pco_hist_ff.describe()

In [ ]:
# Merge output dataset with historical forward fill dataset
pco_hist_ff_merged = pco_hist_ff.merge(right=pco_output,on=['id'])
pco_hist_ff_merged.head()

## **The Main Dataset**

### **Visualizing the amount of missingness**

In [ ]:
# Visualize the completeness of the dataframe
msno.bar(pco_main)

In [ ]:
# Visualize the locations of the missing values of the dataset
msno.matrix(pco_main)

In [ ]:
sorted_main = pco_main.sort_values('date_first_activ')
msno.matrix(sorted_main)

In [ ]:
msno.heatmap(pco_main)

In [ ]:
# Demonstrate why the date_activ column cannot replace completely date_first_activ
activity = ['date_activ','date_first_activ']

# Filter the columns of interest
pco_activity = pco_main[activity]

# Obtain only the complete cases
pco_activity_cc = pco_activity.dropna(subset=['date_first_activ'],how='any',inplace=False)

# Test whether two objects contain the same elements.
pco_activity_cc.date_activ.equals(pco_activity_cc.date_first_activ)

# Describe it
pco_activity_cc.describe(datetime_is_numeric=True) # Comparing dates in .describe() is depricated, might crash in the future

In [ ]:
# Drop the column activity_new and campaign_disc_elec
pco_main_drop = pco_main.drop(labels= ['activity_new','campaign_disc_ele'] , axis=1)

# Remove date_end date_modif_prod date_renewal origin_up pow_max margin_gross_pow_ele margin_net_pow_ele net_margin
brush = ['date_end','date_modif_prod','date_renewal','origin_up','pow_max','margin_gross_pow_ele',
         'margin_net_pow_ele', 'net_margin','forecast_discount_energy','forecast_price_energy_p1',
         'forecast_price_energy_p2','forecast_price_pow_p1']
pco_main_drop.dropna(subset=brush, how='any',inplace=True)

msno.matrix(pco_main_drop)



# **Observations**
* The variable `activity_new` is **MCAR** and has very low correlation with any of the variables. We can safely *drop* this column. 
* The variable `campaign_disc_elec` is completely missing at random on all rows. We can get rid of this column. This suggests that subscribers are not subscribing through campaings offers. 
* The variable `date_first_activ` cannot be replace by the values of the `date_activ` variable. **MAR**
* `net_margin` is showing strong correlation between `margin_gross_pow_elec` and `margin_net_pow_ele`. Multicolinearity is likely here. 
* The variables `origin_up` and `pow_max` display no correlation with any variable and contain 0.54% and 0.01% of missingness respectively. These are **MCAR** and can be dropped listwise.
* `Forecast_base_bill_ele`, `forecast_base_bill_year`, `forecast_bill_12m` and `forecast_cons` variables are highly correlated with the `date_first_activ` variable's missingness. Accouting for 78% of missing values in the formerly mention columns and therfore are **MNAR**.
*Cannot replace the `date_first_activ` column with the `date_activ` column since in some of the cases the dates are not identical. 

In [ ]:
# Choose the columns without missing values
incomplete_cols = ['channel_sales','date_first_activ','forecast_base_bill_ele','forecast_base_bill_year','forecast_bill_12m','forecast_cons']

complete_cols = [column_name for column_name in pco_main_drop.columns if column_name not in incomplete_cols]

pco_main_cc = pco_main_drop[complete_cols]

# Fix negative numeric variables
numeric = [column_name for column_name in pco_main_cc.columns
           if pco_main_cc[column_name].dtype == 'float64' 
           or pco_main_cc[column_name].dtype == 'int64']

# Overwrite positive values on negative values
pco_main_cc[numeric] = pco_main_cc[numeric].apply(abs)

# Describe
pco_main_cc.describe()


In [ ]:
# Convert the has_gas column to  Yes/No
pco_main_cc['has_gas'] = pco_main_cc['has_gas'].replace({'t':'Yes','f':'No'})

# Merge the main dataset with the output dataset
pco_main_cc_merged = pco_main_cc.merge(right=pco_output,on=['id'])

# Convert the churn column to Churned/Stayed
pco_main_cc_merged['churn'] = pco_main_cc_merged['churn'].replace({1:'Churned',0:'Stayed'})

In [ ]:
pco_main_cc_merged.head()

In [ ]:
# Obtain all the variables except for id
variables = [column_name for column_name in pco_main_cc_merged.columns if column_name != 'id']

# Obtain all the categorical variables except for id
categorical = [column_name for column_name in variables if pco_main_cc_merged[column_name].dtype == 'object']

# Obtain all the Date Variables
dates = [column_name for column_name in variables if pco_main_cc_merged[column_name].dtype == 'datetime64[ns]']

# Obtain all the numeric columns
numeric = [column_name for column_name in variables
           if column_name not in categorical 
           and column_name != 'id'
           and column_name != 'churn'
           and column_name not in dates]


# **Data Visualization**

Let's visualize what we've found.

## **The Output Dataset**

In [ ]:
# Calculate the zcores of tenure
tenure_zcores = zscore(a=pco_main_cc_merged['num_years_antig'])
# Convert to absolute values
abs_tenure_zscores = np.abs(tenure_zcores)
# Extract Columns of interest
churn_tenure = pco_main_cc_merged[['churn','num_years_antig']]
# Add z-score column
churn_tenure['z_score'] = list(abs_tenure_zscores)
# Remove outliers 
churned_tenure_filtered = churn_tenure[churn_tenure['z_score'] < 3]
# Visualize tenure by retained customer and churner
vio = sns.violinplot( y=churned_tenure_filtered["churn"], x=churned_tenure_filtered["num_years_antig"] )
# Settings
vio.set(xlabel='Years', ylabel='')
vio.set_title("Customer Attrition by Tenure")
plt.show()

# **Facts**
* The median age of churners is 4 years
* Customers are more likely to churn during the 4th year than the 7th year
* The median age of retained customers is 5 years


# **The Main Dataset**

In [ ]:
# Most popular electricty campaign
ele_nm = pco_main_cc_merged.loc[(pco_main_cc_merged['churn']>='Stayed') & (pco_main_cc_merged['net_margin']>0),['id', 'origin_up','net_margin']]

ele_nm.value_counts(subset=['origin_up'])

In [ ]:
# Highest netting electricity subscription campaign
print(ele_nm.groupby('origin_up')['net_margin'].agg('sum').sort_values(ascending=False))

### **Facts**
* The most popular electricity campaign is `lxidpiddsbxsbosboudacockeimpuepw` which has brought 6,584 current customers.
* The electricity campaign attributable to the highest total net margin is `lxidpiddsbxsbosboudacockeimpuepw`. Netting $1,541,159.95 in 2015.

## **Caveats**


In [ ]:
# Select current customers with positive net margins
top_customers = pco_main_cc_merged.loc[(pco_main_cc_merged['churn']>='Stayed') & (pco_main_cc_merged['net_margin']>0),['id','num_years_antig','net_margin']]

# Top 10 customers by net margin
top_customers.sort_values(by=['net_margin'],ascending=False).head(10)

These are the most profitable customers for PowerCo in terms of net margin. Beware most of them are within the likely tenure of attrition. Time for a marketing campaign!